### DISCUSSION TITLE
debug skills for ocaml

### DISCUSSION THREAD

**Discussion by wanghuibin0** [rems-project/sail#770] [General]
I'm going to working on sail project, but unfamiliar with ocaml.
I feel that ocamldebug is somewhat difficult to use.
Is there any other useful tools for debugging ocaml, or some debugging skills for ocaml?
What's your experience?

**Comment by Alasdair** [comment]
I don't have much experience with ocamldebug

My main approach to triaging bugs in Sail is to write test cases that target specific code paths, and then using the debugging flags/attributes we have to get detailed information on the rewrite and compilation passes. For example if I'm trying to fix a rewrite I'll use the `--ddump-rewrite-ast` flag to identify what is going on. If it's a low-level problem in the C backend individual functions can be tagged with the `$[jib_debug]` attribute which will print a lot of information about the compilation of that specific function.

**Reply by wanghuibin0** [reply]
Thank you for your reply.
I feel that ocamldebug is not very convenient to use. Currently, I primarily use print statements to debug OCaml programs, which requires having well-structured logging information in the code.
Anyway, your experience is useful for me.

**Comment by Alasdair** [comment]
As of https://github.com/rems-project/sail/commit/725cbde6b0140e0ff2b32bf0089f20267fb5f7fc ocamldebug should now work
